src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -88,7 +88,10 @@ private bool EnqueueCore(TElement value)
                 waiter = _waiters.Dequeue();
             }
 
-            waiter.SetResult(value);
+            // Invoke SetResult on a separate task, as this invocation could cause the underlying task to executing,
+            // which could be a long running operation that can potentially cause a deadlock if executed on the current thread.
+            Task.Run(() => waiter.SetResult(value));
```

Comment:
What contract is being used? If it's a new one that has it, you can instead construct the TCS with the RunContinuationsAsynchronously option, which will ensure continuations are not run synchronously.

If you can't do that, for efficiency reasons you could avoid the closure here by using StartNew and passing in the waiter as the object state.

Also earlier in the method, it's be better not to allocate the TCS unless you know you're going to need it.